nuclear/cmake/Modules/GenerateNeutron.cmake

```diff
@@ -0,0 +1,170 @@
+define_property(
+  TARGET
+  PROPERTY NEUTRON_CPP_SOURCE
+  BRIEF_DOCS "Generated neutron C++ source files"
+  FULL_DOCS "Generated neutron C++ source files"
+)
+define_property(
+  TARGET
+  PROPERTY NEUTRON_PROTOBUF_SOURCE
+  BRIEF_DOCS "Generated protobuf C/C++ source files"
+  FULL_DOCS "Generated protobuf C/C++ source files"
+)
+define_property(
+  TARGET
+  PROPERTY NEUTRON_PYTHON_SOURCE
+  BRIEF_DOCS "Generated protobuf python source files"
+  FULL_DOCS "Generated protobuf python source files"
+)
+
+include(CMakeParseArguments)
+function(GenerateNeutron)
+  # We need protobuf and python to generate the neutron messages
+  find_package(Protobuf REQUIRED)
+  find_package(PythonInterp 3 REQUIRED)
+
+  # Set the path to our generating scripts
+  set(SCRIPT_SOURCE "${PROJECT_SOURCE_DIR}/nuclear/cmake/Scripts")
+
+  # Files that are used to generate the neutron files
+  file(GLOB_RECURSE message_class_generator_files "${SCRIPT_SOURCE}/generator/**.py")
+
+  # Extract the arguments from our function call
+  set(options, "")
+  set(oneValueArgs "PROTO" "PARENT_DIR" "BUILTIN_DIR" "BUILTIN_OUTPUT_DIR")
+  set(multiValueArgs "")
+  cmake_parse_arguments(NEUTRON "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  # Locations to store each of the output components
+  set(pb_out "${CMAKE_CURRENT_BINARY_DIR}/protobuf")
+  set(nt_out "${CMAKE_CURRENT_BINARY_DIR}/neutron")
+  set(py_out "${CMAKE_CURRENT_BINARY_DIR}/python")
+
+  # Extract the components of the filename that we need
+  get_filename_component(file_we ${NEUTRON_PROTO} NAME_WE)
```

Comment:
:confused: `we`? I think that this isn't a good name, it does not really describe its job clearly.

Reply:
file whatever? describes its purpose extremely well i think :stuck_out_tongue: 

Also given that the cmake option is called `NAME_WE` i think its hard to get more explicit?

Reply:
I must not know enough about cmake :shrug: